internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -100,159 +102,204 @@ func (r *Reconciler) getNewMachineSet(ctx context.Context, d *clusterv1.MachineD
-	// annotationsToSkip from the parent deployment, and update revision, desiredReplicas,
-	// and maxReplicas) and also update the revision annotation in the deployment with the
-	// latest revision.
+	// Latest MachineSet exists. We need to sync all the in-place propagated values.
 	if existingNewMS != nil {
-		msCopy := existingNewMS.DeepCopy()
-		patchHelper, err := patch.NewHelper(msCopy, r.Client)
-		if err != nil {
-			return nil, err
-		}
-
-		// Set existing new machine set's annotation
-		annotationsUpdated := mdutil.SetNewMachineSetAnnotations(d, msCopy, newRevision, true, log)
-
-		minReadySecondsNeedsUpdate := msCopy.Spec.MinReadySeconds != *d.Spec.MinReadySeconds
-		deletePolicyNeedsUpdate := d.Spec.Strategy.RollingUpdate.DeletePolicy != nil && msCopy.Spec.DeletePolicy != *d.Spec.Strategy.RollingUpdate.DeletePolicy
-		if annotationsUpdated || minReadySecondsNeedsUpdate || deletePolicyNeedsUpdate {
-			msCopy.Spec.MinReadySeconds = *d.Spec.MinReadySeconds
-
-			if deletePolicyNeedsUpdate {
-				msCopy.Spec.DeletePolicy = *d.Spec.Strategy.RollingUpdate.DeletePolicy
+		// FIXME: In the current implementation any changes to in-place propagated fields do not cause the revision to change.
+		// Therefore we loose track of the history of MachineDeployment changes and therefore it wont be able to revert back
+		// to an old state (a state that differs only in the in-place propagated field values).
+		// In the next iteration this will be changed so that in-place propagated values are also captures in history while
+		// not triggering a rollout.
+
+		// The existingMS's revision should be the greatest among all MSes. Usually, its revision number is newRevision
+		// (the max revision number of all old MSes + 1). However, it's possible that some old MSes are deleted
+		// after the existingMS's revision was updated, and newRevision becomes smaller than the current existingMS's
+		// revision. We should only update existingMS's revision when it's smaller than newRevision.
+		// TODO: check if the above comment is actually correct. (The comment was copied from old code). Can the case actually happen?
+		oldRevision, ok := existingNewMS.Annotations[clusterv1.RevisionAnnotation]
+		if ok {
+			oldRevisionInt, err := strconv.ParseInt(oldRevision, 10, 64)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to parse oldRevision on MachineSet")
+			}
+			if newRevisionInt < oldRevisionInt {
+				newRevision = oldRevision
 			}
+		}
 
-			return nil, patchHelper.Patch(ctx, msCopy)
+		desiredMS, err := r.computeDesiredMachineSet(d, existingNewMS, oldMSs, newRevision)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to compute desired MachineSet %q", klog.KObj(existingNewMS))
 		}
+		patchOptions := []client.PatchOption{
+			client.ForceOwnership,
+			client.FieldOwner(machineDeploymentManagerName),
+		}
+		if err = r.Client.Patch(ctx, desiredMS, client.Apply, patchOptions...); err != nil {
```

Comment:
What we are doing here is continuously reconcile not only the fields that are propagated in place but the entire MachineSet. It is kind of overkill, but I won't block on it if we make sure that we are not taking ownership of other's fields.

Reply:
To be honest I don't see an alternative.

I think the option you are bringing up is writing one partial MachineSet with Create and then labels/annotations with SSA.

The initial Create can't contain any labels/annotations though because otherwise we end up with co-ownership once SSA reconciled labels/annotations.

Am I missing something?

Reply:
I agree, the alternative seems worst than this solution, but given that in the prosal we discussed upgrading only some fields in places we should make it the new approach easy to understand to reviewers (explain why we are doing this) and take care of ensuring that this doesn't introduce unexpected side effects